Mathlib/CategoryTheory/Limits/Shapes/Types.lean

```diff
@@ -0,0 +1,623 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+
+! This file was ported from Lean 3 source module category_theory.limits.shapes.types
+! leanprover-community/mathlib commit 70fd9563a21e7b963887c9360bd29b2393e6225a
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.CategoryTheory.Limits.Types
+import Mathlib.CategoryTheory.Limits.Shapes.Products
+import Mathlib.CategoryTheory.Limits.Shapes.BinaryProducts
+import Mathlib.CategoryTheory.Limits.Shapes.Terminal
+import Mathlib.CategoryTheory.ConcreteCategory.Basic
+import Mathlib.Tactic.Elementwise
+
+/-!
+# Special shapes for limits in `Type`.
+
+The general shape (co)limits defined in `CategoryTheory.Limits.Types`
+are intended for use through the limits API,
+and the actual implementation should mostly be considered "sealed".
+
+In this file, we provide definitions of the "standard" special shapes of limits in `Type`,
+giving the expected definitional implementation:
+* the terminal object is `PUnit`
+* the binary product of `X` and `Y` is `X × Y`
+* the product of a family `f : J → Type` is `Π j, f j`
+* the coproduct of a family `f : J → Type` is `Σ j, f j`
+* the binary coproduct of `X` and `Y` is the sum type `X ⊕ Y`
+* the equalizer of a pair of maps `(g, h)` is the subtype `{x : Y // g x = h x}`
+* the coequalizer of a pair of maps `(f, g)` is the quotient of `Y` by `∀ x : Y, f x ~ g x`
+* the pullback of `f : X ⟶ Z` and `g : Y ⟶ Z` is the subtype `{ p : X × Y // f p.1 = g p.2 }`
+  of the product
+
+We first construct terms of `IsLimit` and `LimitCone`, and then provide isomorphisms with the
+types generated by the `HasLimit` API.
+
+As an example, when setting up the monoidal category structure on `Type`
+we use the `types_has_terminal` and `types_has_binary_products` instances.
+-/
+
+
+universe u v
+
+open CategoryTheory
+
+open CategoryTheory.Limits
+
+namespace CategoryTheory.Limits.Types
+
+--attribute [local tidy] tactic.discrete_cases
+
+/-- A restatement of `Types.Limit.lift_π_apply` that uses `Pi.π` and `Pi.lift`. -/
+@[simp 1001]
+theorem pi_lift_π_apply {β : Type u} (f : β → Type u) {P : Type u} (s : ∀ b, P ⟶ f b) (b : β)
+    (x : P) : (Pi.π f b : (∏ f) → f b) (@Pi.lift β _ _ f _ P s x) = s b x :=
+  congr_fun (limit.lift_π (Fan.mk P s) ⟨b⟩) x
+#align category_theory.limits.types.pi_lift_π_apply CategoryTheory.Limits.Types.pi_lift_π_apply
+
+/-- A restatement of `Types.Limit.map_π_apply` that uses `Pi.π` and `Pi.map`. -/
+@[simp 1001]
+theorem pi_map_π_apply {β : Type u} {f g : β → Type u} (α : ∀ j, f j ⟶ g j) (b : β) (x) :
+    (Pi.π g b : (∏ g) → g b) (Pi.map α x) = α b ((Pi.π f b : (∏ f) → f b) x) :=
+  Limit.map_π_apply' _ _ _
+#align category_theory.limits.types.pi_map_π_apply CategoryTheory.Limits.Types.pi_map_π_apply
+
+/-- The category of types has `PUnit` as a terminal object. -/
+def terminalLimitCone : Limits.LimitCone (Functor.empty (Type u))
```

Comment:
Should have a porting note about the regression without `tidy`, etc.

Reply:
All done.